luma_core/src/ios.rs

```diff
@@ -0,0 +1,225 @@
+//! ``ios`` module of ``luma_core``.
+//!
+//! Contains functions for access to the Starlet when running IOS
+
+use crate::cache::DCFlushRange;
+use crate::io::{read32, write32};
+use alloc::boxed::Box;
+
+const BASE: u32 = 0x0d000000;
+
+const HW_IPC_PPCMSG: u32 = BASE + 0;
+const HW_IPC_PPCCTRL: u32 = BASE + 4;
+const HW_IPC_ARMMSG: u32 = BASE + 8;
+
+/// The type of a file descriptor in IOS.
+pub type RawFd = i32;
+
+/// How to open a file.
+#[repr(i32)]
+pub enum Mode {
+    /// With no read/write access.
+    None = 0,
+
+    /// With read only access.
+    Read = 1,
+
+    /// With write only access.
+    Write = 2,
+
+    /// With read/write access.
+    ReadWrite = 3,
+}
+
+/// This is copied from std::io::SeekFrom, with the inner types changed to accomodate IOS.
+#[derive(Copy, PartialEq, Eq, Clone, Debug)]
+pub enum SeekFrom {
+    /// From the start of the file.
+    Start(i32),
+
+    /// From the current position.
+    Current(i32),
+
+    /// From the end of the file.
+    End(i32),
+}
+
+#[derive(Debug, Clone, PartialEq)]
+#[repr(u32)]
+enum Command {
+    Open = 1,
+    Close = 2,
+    Read = 3,
+    Write = 4,
+    Seek = 5,
+    Ioctl = 6,
+    Ioctlv = 7,
+    Async = 8,
+}
+
+#[derive(Debug, Clone)]
+#[repr(C, align(32))]
+struct Ipc {
+    command: Command,
+    ret: i32,
+    fd: RawFd,
+    args: [i32; 5],
+}
+
+impl Ipc {
+    #[inline]
+    fn new(command: Command, args: [i32; 5]) -> Box<Self> {
+        Box::new(Ipc {
+            command,
+            ret: 0,
+            fd: -1,
+            args,
+        })
+    }
+
+    #[inline]
+    fn with_fd(command: Command, fd: i32, args: [i32; 5]) -> Box<Self> {
+        Box::new(Ipc {
+            command,
+            ret: 0,
+            fd,
+            args,
+        })
+    }
+
+    #[inline(never)]
+    fn send(self: Box<Self>) -> Box<Self> {
+        let ptr = Box::into_raw(self);
+
+        // Flush the IPC data from its cache line, so that the Starlet will see it.
+        unsafe { DCFlushRange(ptr as *const _, core::mem::size_of::<Self>() as u32) };
```

Comment:
you can use `try_into` here instead of  just casting to a `u32`. maybe using a `usize` which is the same because of `target_ptr_width = 32` could be an option